# Data Keystone: Exploring the Dataset

## Loading the Dataset

In [16]:
import pandas as pd

df = pd.read_csv('steam_games_dataset.csv')

## Exploring the Dataset Structure

In [17]:
print(df.head())
print(df.info())
print(df.shape)
df.columns
df.dtypes

     appid                              name         developer  \
0      730  Counter-Strike: Global Offensive             Valve   
1  1172470                      Apex Legends           Respawn   
2   578080               PUBG: BATTLEGROUNDS  PUBG Corporation   
3  1623730                          Palworld        Pocketpair   
4      440                   Team Fortress 2             Valve   

         publisher  score_rank  positive  negative  userscore  \
0            Valve         NaN   7642084   1173003          0   
1  Electronic Arts         NaN    668053    326926          0   
2    KRAFTON, Inc.         NaN   1520457   1037487          0   
3       Pocketpair         NaN    358266     22443          0   
4            Valve         NaN   1044264    117208          0   

                       owners  average_forever  average_2weeks  \
0  100,000,000 .. 200,000,000            33852             708   
1  100,000,000 .. 200,000,000            10506             496   
2  100,000,000

appid                int64
name                   str
developer              str
publisher              str
score_rank         float64
positive             int64
negative             int64
userscore            int64
owners                 str
average_forever      int64
average_2weeks       int64
median_forever       int64
median_2weeks        int64
price                int64
initialprice         int64
discount             int64
ccu                  int64
dtype: object

## Exploring Column Values

In [18]:
print(df.value_counts())
print(df.describe())

appid   name                            developer                 publisher                score_rank  positive  negative  userscore  owners                average_forever  average_2weeks  median_forever  median_2weeks  price  initialprice  discount  ccu
22490   Fallout: New Vegas              Obsidian Entertainment    Bethesda Softworks       100.0       46753     2153      95         500,000 .. 1,000,000  3333             379             1409            310            999    999           0         318    1
790650  MARVEL END TIME ARENA           Smilegate Entertainment   Smilegate Entertainment  99.0        368       160       69         100,000 .. 200,000    158              0               158             0              0      0             0         0      1
907680  Wwbit                           1bit                      1bit                     98.0        710       566       55         100,000 .. 200,000    23               0               19              0              0    

In [19]:
df['appid'].max()

np.int64(3605460)

In [20]:
df = df.astype({'appid': 'int32'})
print(df.dtypes)

appid                int32
name                   str
developer              str
publisher              str
score_rank         float64
positive             int64
negative             int64
userscore            int64
owners                 str
average_forever      int64
average_2weeks       int64
median_forever       int64
median_2weeks        int64
price                int64
initialprice         int64
discount             int64
ccu                  int64
dtype: object


In [21]:
df.head()
df.describe()

,appid,score_rank,positive,negative,userscore,average_forever,average_2weeks,median_forever,median_2weeks,price,initialprice,discount,ccu
count,1.000000e+04,5.0,1.000000e+04,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,1.000000e+04
mean,9.008700e+05,99.0,1.214202e+04,1.948096e+03,0.036000,1023.689500,112.496300,515.001700,118.949000,1131.326700,1284.141500,8.675800,5.643178e+02
std,7.290494e+05,1.0,9.536926e+04,1.812977e+04,1.671581,4193.287962,638.473189,3508.429257,707.962801,1299.903805,1364.327698,23.476133,1.128781e+04
min,1.000000e+01,98.0,0.000000e+00,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00
25%,3.365875e+05,98.0,4.310000e+02,1.100000e+02,0.000000,145.000000,0.000000,79.750000,0.000000,99.000000,199.000000,0.000000,0.000000e+00
50%,6.551300e+05,99.0,1.514500e+03,2.950000e+02,0.000000,310.000000,0.000000,210.000000,0.000000,719.000000,999.000000,0.000000,5.000000e+00
75%,1.333238e+06,100.0,4.996750e+03,8.712500e+02,0.000000,817.250000,6.000000,406.000000,6.000000,1799.000000,1999.000000,0.000000,4.500000e+01
max,3.605460e+06,100.0,7.642084e+06,1.173003e+06,95.000000,263340.000000,20080.000000,269193.000000,20080.000000,14999.000000,14999.000000,95.000000,1.013936e+06


**score_rank** only has 5 values, which has potential to hinder a full picture of user experience feedback. This field i worth investigating further.

In [22]:
df.head(15)

,appid,name,developer,publisher,score_rank,positive,negative,userscore,owners,average_forever,average_2weeks,median_forever,median_2weeks,price,initialprice,discount,ccu
0,730,Counter-Strike: Global Offensive,Valve,Valve,NaN,7642084,1173003,0,"100,000,000 .. 200,000,000",33852,708,6645,301,0,0,0,1013936
1,1172470,Apex Legends,Respawn,Electronic Arts,NaN,668053,326926,0,"100,000,000 .. 200,000,000",10506,496,935,246,0,0,0,124262
2,578080,PUBG: BATTLEGROUNDS,PUBG Corporation,"KRAFTON, Inc.",NaN,1520457,1037487,0,"100,000,000 .. 200,000,000",23165,717,5622,261,0,0,0,314682
3,1623730,Palworld,Pocketpair,Pocketpair,NaN,358266,22443,0,"50,000,000 .. 100,000,000",3854,835,2213,257,2999,2999,0,18028
4,440,Team Fortress 2,Valve,Valve,NaN,1044264,117208,0,"50,000,000 .. 100,000,000",21244,736,4262,102,0,0,0,43819
5,1938090,Call of Duty: Modern Warfare II,"Treyarch, Raven Software, Beenox, High Moon St...",Activision,NaN,419594,294520,0,"50,000,000 .. 100,000,000",5397,639,738,324,3849,6999,45,67419
6,1063730,New World: Aeternum,Amazon Games,Amazon Games,NaN,196798,90080,0,"50,000,000 .. 100,000,000",10588,18,3762,18,5999,5999,0,5583
7,2358720,Black Myth: Wukong,Game Science,Game Science,NaN,1111720,38378,0,"50,000,000 .. 100,000,000",3268,524,2838,202,5999,5999,0,15004
8,271590,Grand Theft Auto V Legacy,Rockstar North,Rockstar Games,NaN,1739980,250576,0,"50,000,000 .. 100,000,000",13153,699,5454,49,0,0,0,67851
9,550,Left 4 Dead 2,Valve,Valve,NaN,940221,23762,0,"50,000,000 .. 100,000,000",2479,352,543,71,999,999,0,25122


The first 15 records have a **score_rank** value of NaN. Is there a way to break this down and rebuild the ranking?

## Practicing Basic Cleaning and Wrangling

In [23]:
df[df['score_rank'].isna() == False]

,appid,name,developer,publisher,score_rank,positive,negative,userscore,owners,average_forever,average_2weeks,median_forever,median_2weeks,price,initialprice,discount,ccu
2122,22490,Fallout: New Vegas,Obsidian Entertainment,Bethesda Softworks,100.0,46753,2153,95,"500,000 .. 1,000,000",3333,379,1409,310,999,999,0,318
6754,790650,MARVEL END TIME ARENA,Smilegate Entertainment,Smilegate Entertainment,99.0,368,160,69,"100,000 .. 200,000",158,0,158,0,0,0,0,0
6960,907680,Wwbit,1bit,1bit,98.0,710,566,55,"100,000 .. 200,000",23,0,19,0,0,0,0,0
8680,823550,Booty Calls,3X Entertainment Limited,Nutaku Publishing,98.0,255,294,46,"100,000 .. 200,000",272,0,131,0,0,0,0,33
9662,639780,Deep Space Waifu: FLAT JUSTICE,Neko Climax Studios,Neko Climax Studios,100.0,1449,67,95,"50,000 .. 100,000",137,0,111,0,399,399,0,0


The 5 records with **score_rank** having a numerical value also have a **userscore** that is greater than 0. Compared to the previous data sample, all records with **score_rank** of NaN have a **userscore** of 0.  

Additionally, the relationship between **score_rank**, **userscore**, **positive**, and **negative** seem illogical at initial glance. Record 2122 and Record 9662 both share a userscore value of 95, suggesting why they both share a **score_rank** of 100.0. This also presents that a higher **score_rank** reflects positively rather than a lower value, so this field may be measured in a percentile-sense. Renaming the column for clarity may be necessary. 

I have suspicion of the scoring, but I want to see if we can rebuild that column. Let's make a dataframe of the first several fields and include just the records with numerical **score_rank** values.

In [24]:
df_scored = df[df['score_rank'].notna()][['appid', 'name', 'score_rank', 'positive', 'negative', 'userscore']]
df_scored['positive_ratio'] = (df['positive'] / (df['positive'] + df['negative'])) * 100

df_scored.head()

,appid,name,score_rank,positive,negative,userscore,positive_ratio
2122,22490,Fallout: New Vegas,100.0,46753,2153,95,95.597677
6754,790650,MARVEL END TIME ARENA,99.0,368,160,69,69.696970
6960,907680,Wwbit,98.0,710,566,55,55.642633
8680,823550,Booty Calls,98.0,255,294,46,46.448087
9662,639780,Deep Space Waifu: FLAT JUSTICE,100.0,1449,67,95,95.580475


This confirms that this field's values come from a simple ratio, but it is of note that this data has rounded down all values. I'm not sure why that would be the case. With this information, we should be able to calculate and fill all remaining missing values for the **userscore** field as well as fill and sort the **score_rank** field.

In [27]:
df['positive_ratio'] = ((df['positive'] / (df['positive'] + df['negative'])) * 100).round(0).astype('Int64')
df['positive_ratio'].head(15)

0     87
1     67
2     59
3     94
4     90
5     59
6     69
7     97
8     87
9     98
10    91
11    71
12    64
13    58
14    76
Name: positive_ratio, dtype: Int64

In [28]:
df.head()

,appid,name,developer,publisher,score_rank,positive,negative,userscore,owners,average_forever,average_2weeks,median_forever,median_2weeks,price,initialprice,discount,ccu,positive_ratio
0,730,Counter-Strike: Global Offensive,Valve,Valve,NaN,7642084,1173003,0,"100,000,000 .. 200,000,000",33852,708,6645,301,0,0,0,1013936,87
1,1172470,Apex Legends,Respawn,Electronic Arts,NaN,668053,326926,0,"100,000,000 .. 200,000,000",10506,496,935,246,0,0,0,124262,67
2,578080,PUBG: BATTLEGROUNDS,PUBG Corporation,"KRAFTON, Inc.",NaN,1520457,1037487,0,"100,000,000 .. 200,000,000",23165,717,5622,261,0,0,0,314682,59
3,1623730,Palworld,Pocketpair,Pocketpair,NaN,358266,22443,0,"50,000,000 .. 100,000,000",3854,835,2213,257,2999,2999,0,18028,94
4,440,Team Fortress 2,Valve,Valve,NaN,1044264,117208,0,"50,000,000 .. 100,000,000",21244,736,4262,102,0,0,0,43819,90


The **positive_ratio** field has been added to the end. Now we just need to reorder the field positions to group the relevant data together.

In [29]:
cols = df.columns.tolist()
cols.remove('positive_ratio')
cols.insert(8, 'positive_ratio')
df = df[cols]
df.head()

,appid,name,developer,publisher,score_rank,positive,negative,userscore,positive_ratio,owners,average_forever,average_2weeks,median_forever,median_2weeks,price,initialprice,discount,ccu
0,730,Counter-Strike: Global Offensive,Valve,Valve,NaN,7642084,1173003,0,87,"100,000,000 .. 200,000,000",33852,708,6645,301,0,0,0,1013936
1,1172470,Apex Legends,Respawn,Electronic Arts,NaN,668053,326926,0,67,"100,000,000 .. 200,000,000",10506,496,935,246,0,0,0,124262
2,578080,PUBG: BATTLEGROUNDS,PUBG Corporation,"KRAFTON, Inc.",NaN,1520457,1037487,0,59,"100,000,000 .. 200,000,000",23165,717,5622,261,0,0,0,314682
3,1623730,Palworld,Pocketpair,Pocketpair,NaN,358266,22443,0,94,"50,000,000 .. 100,000,000",3854,835,2213,257,2999,2999,0,18028
4,440,Team Fortress 2,Valve,Valve,NaN,1044264,117208,0,90,"50,000,000 .. 100,000,000",21244,736,4262,102,0,0,0,43819


In [35]:
print(df['positive_ratio'].value_counts()[0])
print(df['userscore'].value_counts()[0])


25
9995


We have reduced the 0 values from 9,995 counts in our 10,000 records within the **userscore** field to just 25 counts within the **positive_ratio** field, a substantial improvement over the source-available data.